File: q51-75/q53.py
# 2020.08.21
# Problem Statement:
# https://leetcode.com/problems/maximum-subarray/

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        # idea:
        # when partial sum < 0, do not hesitate to remove that part, 
        # ie: start a new discovery.

        # check empty case
        if len(nums) == 0: return 0

        # initialize answer as the max element in nums
        answer = max(nums)
        
        # initialize left bound and right bound of sum
        left, right = 0, 0
        current_sum = nums[0]

        # when not reach the end yet
        while right <= len(nums)-1:
            # have reached the end of list, return the max between current result and previous max
            if right == len(nums)-1: return max(answer, current_sum)
            # remove that part and start new
            elif current_sum < 0:
                left, right = right + 1, right + 1
                current_sum = nums[left]
            # include the next element
            else:
                right = right + 1
                current_sum = current_sum + nums[right]
                if nums[right] > 0:
                    answer = max(answer, current_sum)
                    

File: Amazon/q53.java
// 2022.05.26 midnight
// Problem Statement:
// https://leetcode.com/problems/maximum-subarray/

// idea: move start and end continuously, if the sum of the subarray [start, end] <= 0,
// abandon this subarray and move both start and end to the next,
// otherwise, only move end forward
class Solution {
    public int maxSubArray(int[] nums) {
        int start = 0, end = 0;
        int answer = Integer.MIN_VALUE;
        int curr_sum = 0;
        while (end!=nums.length) {
            curr_sum += nums[end];
            if (curr_sum<=0) {
                answer = Math.max(answer, curr_sum);
                curr_sum = 0;
                start = end+1;
                end++;
            } else {
                end++;
                answer = Math.max(answer, curr_sum);
            }
        }
        return answer;
    }
}